src/hash_table_internal.h

```diff
@@ -46,32 +43,67 @@
  */
 struct ly_ht_rec {
     uint32_t hash;        /* hash of the value */
-    int32_t hits;         /* collision/overflow value count - 1 (a filled entry has 1 hit,
-                           * special value -1 means a deleted record) */
-    unsigned char val[1]; /* arbitrary-size value */
-} _PACKED;
+    uint32_t next;        /* index of next collision */
+    unsigned char val[]; /* arbitrary-size value */
```

Comment:
When writing this, I researched it quite extensively but I do not remember the details. Bu I think that, strictly according to the standard, this is not a valid declaration and some compilers may have a problem with it.

Reply:
This syntax is conform to C99. See "6.7.2.1 Structure and union specifiers" in https://www.open-std.org/JTC1/SC22/WG14/www/docs/n1256.pdf

I know that val[0] is indeed not portable (gcc-ism), but I never had issues with val[]. However I have no experience outside Unix.

Also, I wonder if val[1] couldn't be interpreted by the compiler as a fixed array, which could trigger an out-of-bound access.

Reply:
I think it was recommended as the safe choice that should always work. Also, what is and out-of-bound access? Never heard about such checks in C, it should not care. I am not strictly against it but `[1]` worked for quite a lot of time and no issues were reported whatsoever so I do not think there is any reason to change it.